Docs: Node comparison report (#7346)
ChainSafe/forest: docs/docs/users/reports/node_comparison/index.md

---
sidebar_position: 1

> title: Forest vs Lotus

Reviewer: This shouldn't really be a Forest vs Lotus, just Filecoin Node comparison. Venus should also be included there, at least an extended mention.

---

# Forest vs Lotus

> Forest and Lotus are both full-node implementations of the Filecoin protocol. Forest is written in Rust and maintained by [ChainSafe Systems](https://chainsafe.io); it focuses on chain validation, a high-performance RPC API, and snapshot generation, exporting a chain snapshot substantially faster than Lotus while using a fraction of the memory. Lotus is written in Go; it provides the complete Filecoin feature set, including the storage-provider stack.

Reviewer: ```suggestion
Forest and Lotus are both full-node implementations of the Filecoin protocol. Forest is written in Rust and maintained by [ChainSafe Systems](https://chainsafe.io); it focuses on chain validation, a high-performance RPC API, and snapshot generation with low hardware requirements. Lotus is the reference implementation written in Go; it provides the complete Filecoin feature set, including the storage-provider stack.
```

- Choose **Forest** to validate the chain, serve the Filecoin or Ethereum RPC API, generate snapshots, or run a bootstrap node, particularly where lower resource usage matters.

> - Choose **Lotus** when you need storage-provider functionality (sealing, proving, block production) or the completeness of the reference implementation, where new protocol features typically land first.

Reviewer: ```suggestion
```

Let's not tell people what to do. Rather, we should facilitate making informed decisions.

## Feature comparison

| Capability                                                 | Forest | Lotus |
| ---------------------------------------------------------- | ------ | ----- |
| Chain synchronization and validation                       | Yes    | Yes   |

> | Filecoin JSON-RPC API                                      | Yes    | Yes   |

Reviewer: This needs more details. Where are we compatible? All `/v0`? All `/v1`? All `/v2`?. Otherwise you can just say Filecoin JSON-RPC API with only one method implemented for `/v0`.

| Ethereum-compatible RPC (`eth_*`)                          | Yes    | Yes   |
| Snapshot export                                            | Yes    | Yes   |
| Built-in wallet                                            | Yes    | Yes   |
| Bootstrap node                                             | Yes    | Yes   |
| F3 (Fast Finality) participation                           | Yes    | Yes   |
| Storage provider / sealing (`lotus-miner`, `lotus-worker`) | No     | Yes   |

> | Block production / mining                                  | No     | Yes   |

Reviewer: ```suggestion
| Block production / mining                                  | No (untested)     | Yes   |
```

## Performance

> For a comparable RPC workload, Forest served requests at lower latency while using less CPU and memory than Lotus, and it exported a snapshot in a fraction of the time and memory. See the [RPC Performance Comparison](./rpc_comparison.md) and [Snapshot Generation Comparison](./snapshot_comparison.md) for the full figures and methodology.

Reviewer: ```suggestion
For a comparable RPC workload, Forest served requests at lower latency while using less CPU and memory than Lotus, and it exports a snapshot significantly faster and with lower hardware requirements. See the [RPC Performance Comparison](./rpc_comparison.md) and [Snapshot Generation Comparison](./snapshot_comparison.md) for the full figures and methodology.
```

# Filecoin Node Comparison

There are several full-node implementations of the Filecoin protocol.

- **[Forest](https://github.com/ChainSafe/forest)** is written in Rust and maintained by [ChainSafe Systems](https://chainsafe.io). It focuses on chain validation, a high-performance RPC API, and snapshot generation with low hardware requirements.
- **[Lotus](https://github.com/filecoin-project/lotus)** is the reference implementation, written in Go. It provides the complete Filecoin feature set, including the storage-provider stack, and is typically where new protocol features land first.
- **[Venus](https://github.com/filecoin-project/venus)** is a modular Go implementation. Rather than a single daemon, it is a set of independently deployable components (`venus`, `damocles`, `sophon-miner`, `venus-wallet`, `sophon-messager`, `sophon-auth`, `sophon-gateway`) that together pioneered Filecoin's distributed storage pool model, with a largely Lotus-compatible API. It aims to help small and medium-sized storage providers join the Filecoin network with a lower barrier to entry.

## Feature comparison

| Capability                                                 | Forest        | Lotus |

> | ---------------------------------------------------------- | ------------- | ----- |

Reviewer: Let's add some features that Forest has that Lotus doesn't like eth tracing methods, archival lite/diff snapshots generation, auto-download snapshot on startup, single binary for all networks.

Reviewer: I'm not sure if Lotus has healthcheck endpoint - something to check out

Reviewer: separate wallet

Author: Added

| Chain synchronization and validation                       | Yes           | Yes   |
| Filecoin JSON-RPC API                                      | Yes           | Yes   |
| Ethereum-compatible RPC (`eth_*`)                          | Yes           | Yes   |
| Snapshot export                                            | Yes           | Yes   |
| Built-in wallet                                            | Yes           | Yes   |
| Bootstrap node                                             | Yes           | Yes   |
| F3 (Fast Finality) participation                           | Yes           | Yes   |
| Storage provider / sealing (`lotus-miner`, `lotus-worker`) | No            | Yes   |

> | Block production / mining                                  | No (untested) | Yes   |

Reviewer: ```suggestion
| Block production / mining                                  | No (untested) | Yes   |
| Written in Rust 🦀  | Yes | No |
```
;)

> Forest and Lotus both expose a Lotus-compatible Filecoin JSON-RPC API (requests and responses use the same JSON format). Forest serves three API versions: `/rpc/v0` (deprecated, legacy Lotus-compatible methods), `/rpc/v1` (stable and recommended for production), and `/rpc/v2` (experimental, still being rolled out). Forest does not aim for 100% Lotus API parity: it implements the methods needed for chain validation, RPC serving, and snapshots, but not storage-provider or sealing methods. For the full, per-version list of supported methods, see the [JSON-RPC overview](../../reference/json-rpc/overview.md) and [methods reference](../../reference/json-rpc/methods.mdx).

Reviewer: ```suggestion
Forest and Lotus both expose a compatible Filecoin JSON-RPC API (requests and responses use the same JSON format). Forest serves three API versions: `/rpc/v0` (deprecated, legacy Lotus-compatible methods), `/rpc/v1` (stable and recommended for production), and `/rpc/v2` (experimental, still being rolled out). For the full, per-version list of supported methods, see the [JSON-RPC overview](../../reference/json-rpc/overview.md) and [methods reference](../../reference/json-rpc/methods.mdx).
```

Also, maybe mention https://docs.forest.chainsafe.io/reports/api_conformance/report_2026-07-01

---
sidebar_position: 1
title: Filecoin Node Comparison
---

# Filecoin Node Comparison

> There are several full-node implementations of the Filecoin protocol.

Reviewer: ```suggestion
There are three full-node implementations of the Filecoin protocol.
```

- **[Forest](https://github.com/ChainSafe/forest)** is written in Rust and maintained by [ChainSafe Systems](https://chainsafe.io). It focuses on chain validation, a high-performance RPC API, and snapshot generation, all with low hardware requirements.
- **[Lotus](https://github.com/filecoin-project/lotus)** is the reference implementation, written in Go. It provides the complete Filecoin feature set, including the storage-provider stack, and is typically where new protocol features land first.

> - **[Venus](https://github.com/filecoin-project/venus)** is a modular Go implementation. Rather than a single daemon, it is a set of independently deployable components (`venus`, `damocles`, `sophon-miner`, `venus-wallet`, `sophon-messager`, `sophon-auth`, `sophon-gateway`) that together pioneered Filecoin's distributed storage pool model and offer a largely Lotus-compatible API. It aims to help small and medium-sized storage providers join the Filecoin network with a lower barrier to entry.

Reviewer: ```suggestion
- **[Venus](https://github.com/filecoin-project/venus)** is a modular Go implementation. Rather than a single daemon, it is a set of independently deployable components that together pioneered Filecoin's distributed storage pool model and offer a largely Lotus-compatible API. It aims to help small and medium-sized storage providers join the Filecoin network with a lower barrier to entry.
```

We don't name Forest and Lotus binaries, no need to do that for Venus, no?

| Bootstrap node                                                    | Yes           | Yes   |
| F3 (Fast Finality) participation                                  | Yes           | Yes   |
| Built-in health-check endpoints                                   | Yes           | Yes   |
| Storage provider / sealing (`lotus-miner`, `lotus-worker`)        | No            | Yes   |
| Block production / mining                                         | No (untested) | Yes   |
| Single binary for all networks (`--chain`)                        | Yes           | No    |
| Auto-download snapshot on startup                                 | Yes           | No    |
| Lite / diff / archival snapshot generation                        | Yes           | No    |
| Offline RPC server from snapshot (`forest-tool api serve`)        | Yes           | No    |
| Stateless node mode (`--stateless`)                               | Yes           | No    |
| Ethereum `trace_call` and `debug_traceTransaction`                | Yes           | No    |
| Written in Rust 🦀                                                | Yes           | No    |

> Forest and Lotus both expose a compatible Filecoin JSON-RPC API (requests and responses use the same JSON format). Forest supports three API versions: `/rpc/v0` (deprecated; legacy Lotus-compatible methods), `/rpc/v1` (stable and recommended for production), and `/rpc/v2` (experimental; still being rolled out). See the [JSON-RPC overview](../../reference/json-rpc/overview.md), [methods reference](../../reference/json-rpc/methods.mdx), and [API conformance report](../api_conformance/report_2026-07-01.md) for supported methods and Lotus API parity.

Reviewer: ```suggestion
Forest and Lotus both expose a compatible Filecoin JSON-RPC API. Forest supports three API versions: `/rpc/v0` (deprecated; legacy Lotus-compatible methods), `/rpc/v1` (stable and recommended for production), and `/rpc/v2` (experimental; still being rolled out). See the [JSON-RPC overview](../../reference/json-rpc/overview.md), [methods reference](../../reference/json-rpc/methods.mdx), and [API conformance report](../api_conformance/report_2026-07-01.md) for supported methods and Lotus API parity.
```

Let's assume the reader knows what `compatible` means, no need to explain basic technical nomenclature.
